File: files/CH-122.R
library(tidyverse)
library(readxl)

path = "files/CH-122 Table Transformation.xlsx"
input = read_excel(path, range = "C2:E27")
test  = read_excel(path, range = "G2:J17") %>%
  mutate(Date = format(Date, "%d/%m/%Y"))

block_size = 5

result = input %>%
  mutate(row_id = row_number(),
         block_id = (row_id - 1) %/% block_size + 1) %>%
  summarise(
    date = Date[2],
    region = Date[1],
    fruits = list(Description[3:block_size]),
    values = list(Qty[3:block_size]),
    .by = block_id
  ) %>%
  unnest(c(fruits, values)) %>%
  arrange(date, block_id, desc(values)) %>%
  select(-block_id)

names(result) = names(test)

all.equal(result, test, check.attributes = FALSE)
#> [1] TRUE

File: files/100-199/122/CH-122.py
import pandas as pd

path = "CH-122 Table Transformation.xlsx"
input = pd.read_excel(path, usecols="C:E", skiprows=1, nrows=26)
test = pd.read_excel(path, usecols="G:J", skiprows=1, nrows=15).rename(columns=lambda x: x.replace('.1', ''))

block_size = 5

input['row_id'] = input.index + 1
input['block_id'] = (input['row_id'] - 1) // block_size + 1

result = input.groupby('block_id').apply(
    lambda x: pd.DataFrame({
        'date': [x['Date'].iloc[1]] * (block_size - 2),
        'region': [x['Date'].iloc[0]] * (block_size - 2),
        'fruits': x['Description'].iloc[2:block_size].tolist(),
        'values': x['Qty'].iloc[2:block_size].tolist()
    })
).reset_index(level=1, drop=True)

result = result.sort_values(by=['date', 'block_id', 'values'], ascending=[True, True, False]).reset_index(drop=True)
result = result.assign(date=pd.to_datetime(result['date']), values=result['values'].astype("int64"))

result.columns = test.columns

print(result.equals(test))
# True